src/t8_schemes/t8_scheme_common_cxx.hxx

```diff
@@ -0,0 +1,631 @@
+
+    if (id == 0) {
+      t8_element_first_descendant (elem, elem, level);
+      return;
+    }
+
+    T8_ASSERT (t8_element_level (elem) < level);
+
+    if (t8_element_level (elem) + 1 == level) {
+      T8_ASSERT (id <= (long unsigned int) t8_element_num_children (elem));
+      t8_element_child (elem, id, elem);
+      return;
+    }
+
+    t8_linearidx_t sum_descendants_of_children_before = 0;
+    t8_linearidx_t num_descendants_of_child = 0;
+    int childindex;
+    for (childindex = 0; childindex < t8_element_num_children (elem); childindex++) {
+      t8_element_child (elem, childindex, elem);
+      num_descendants_of_child = t8_element_count_leaves (elem, level);
+      t8_element_parent (elem, elem);
+
+      sum_descendants_of_children_before += num_descendants_of_child;
+      if (sum_descendants_of_children_before > id) {
+        sum_descendants_of_children_before -= num_descendants_of_child;
+        break;
+      }
+    }
+    t8_element_child (elem, childindex, elem);
+    t8_element_init_linear_id_recursive (elem, level, id - sum_descendants_of_children_before);
+  }
+
+  virtual void
+  t8_element_set_linear_id (t8_element_t *elem, int level, t8_linearidx_t id) const override
+  {
+    t8_element_root (elem);
+    t8_element_init_linear_id_recursive (elem, level, id);
+  }
+
+  virtual t8_linearidx_t
+  t8_element_get_linear_id (const t8_element_t *elem, int level) const override
+  {
+    t8_element_t *rec_start;
+    t8_element_new (1, &rec_start);
+    /* Determine desc or anc on level */
+    if (level > t8_element_level (elem)) {
+      t8_element_first_descendant (elem, rec_start, level);
+    }
+    else {
+      t8_element_copy (elem, rec_start);
+      while (t8_element_level (rec_start) > level) {
+        t8_element_parent (rec_start, rec_start);
+      }
+    }
+
+    /* Maybe we can also input p into recursive function and calculate id directly for first desc */
+    t8_linearidx_t id = t8_element_linear_id_recursive (rec_start, 0, t8_element_level (rec_start));
+    T8_ASSERT (id >= 0);
+    t8_element_destroy (1, &rec_start);
+    return id;
+  }
+
+  virtual void
+  t8_element_first_descendant (const t8_element_t *elem, t8_element_t *desc, int level) const override
+  {
+    t8_element_copy (elem, desc);
+    while (t8_element_level (desc) < level) {
+      t8_element_child (desc, 0, desc);
+    }
+  }
+
+  virtual void
+  t8_element_last_descendant (const t8_element_t *elem, t8_element_t *desc, int level) const override
+  {
+    t8_element_copy (elem, desc);
+    while (t8_element_level (desc) < level) {
+      t8_element_child (desc, t8_element_num_children (desc) - 1, desc);
+    }
+  }
+
+  virtual void
+  t8_element_successor (const t8_element_t *elem1, t8_element_t *elem2, int level) const override
+  {
+    T8_ASSERT (level != 0);
+    T8_ASSERT (level == t8_element_level (elem1)); /* TODO: if this is true, do we need level in interface? */
```

Comment:
I don't think that this is true. Maybe @holke can help here. We should be able to compute the successor on a different level.

Reply:
No test currently fails, if we add this assertion to one of the current default schemes. I'd argue that a user should be able to build this functionality of differing levels themselves, by combining either anc or first_descendant with successor.
So my preferred solution is to introduce a scheme interface function `ancestor`, as well as removing the level from the `successor` interface